pyrokinetics/diagnostics/Diagnostics.py

```diff
@@ -0,0 +1,169 @@
+        the Poincare Map.
+
+        This routine may take a while depending on ``nturns`` and on
+        the number of magnetic field lines. The parameter rhostar is
+        required by the flux-tube boundary condition.
+        Available for CGYRO, GENE and GS2 nonlinear simulations
+
+        You need to load the output files of a simulation
+        berore calling this function.
+
+        Parameters
+        ----------
+        xarray: numpy.ndarray, array containing x coordinate of initial
+            field line positions
+        yarray: numpy.ndarray, array containing y coordinate of initial
+            field line positions
+        nturns: int, number of intersection points
+        time: float, time reference
+        rhostar: float, rhostar is needed to set the boundary condition
+                 on the magnetic field line
+
+        Returns
+        -------
+        coordinates: numpy.ndarray, 4D array of shape (2, nturns, len(yarray), len(xarray))
+               containing the x and y coordinates shaped according to the initial
+               field line position. See ``example_poincare.py`` for a simple example.
+
+        Raises
+        ------
+        NotImplementedError: if `gk_code` is not `CGYRO`, `GENE` or `GS2`
+        RuntimeError: in case of linear simulation
+        """
+        if self.pyro.gk_code not in ["CGYRO", "GS2", "GENE"]:
+            raise NotImplementedError(
+                "Poincare map only available for CGYRO, GENE and GS2"
+            )
+        if self.pyro.gk_input.is_linear():
+            raise RuntimeError("Poincare only available for nonlinear runs")
+        apar = self.pyro.gk_output.fields.sel(field="apar").sel(
+            time=time, method="nearest"
+        )
+        kx = apar.kx.values
+        ky = apar.ky.values
+        ntheta = apar.theta.shape[0]
+        nkx = kx.shape[0]
+        nky = ky.shape[0]
+        dkx = kx[1] - kx[0]
+        dky = ky[1]
+        ny = 2 * (nky - 1)
+        nkx0 = nkx + 1 - np.mod(nkx, 2)
+        Lx = 2 * np.pi / dkx
+        Ly = 2 * np.pi / dky
+        xgrid = np.linspace(-Lx / 2, Lx / 2, nkx0)[:nkx]
+        ygrid = np.linspace(-Ly / 2, Ly / 2, ny)
+        xmin = np.min(xgrid)
+        ymin = np.min(ygrid)
+        ymax = np.max(ygrid)
+
+        # Geometrical factors
+        geo = self.pyro.local_geometry
+        nskip = len(geo.theta) // ntheta
+        bmag = np.sqrt((1 / geo.R) ** 2 + geo.b_poloidal**2)
+        bmag = np.roll(bmag[::nskip], ntheta // 2)
+        jacob = geo.jacob * geo.dpsidr * geo.get("bunit_over_b0", 1)
+        jacob = np.roll(jacob[::nskip], ntheta // 2)
+        dq = rhostar * Lx * geo.shat / geo.dpsidr
+        qmin = geo.q - dq / 2
+        fac1 = 2 * np.pi * geo.dpsidr / rhostar
+        fac2 = geo.dpsidr * geo.q / geo.rho
+
+        # Compute bx and by
+        ikxapar = 1j * apar.kx * apar * nkx * ny
+        ikyapar = -1j * apar.ky * apar * nkx * ny
```

Comment:
Should we move the nkx and ky inside the FFT as this isn't really ikxapar/ikyapar anymore?

Reply:
This is done.